Brought back eager based runtime, as a stepping stone towards tensor handle support.

mhong opened the pull request:
Switched CPU and GPU runtime unit tests to eager.

This PR adapts and extends the deleted code from https://github.com/apple/swift/commit/684177fe94a1eb7a03f4f1403cd0473a70f885bc#diff-3d9a6452da404ba617a54693692a94f9.

Disabled a few unit tests for GPU execution, and filed https://bugs.swift.org/browse/SR-8711 to track
them. This PR is however expected to be usable for Jupyter notebook in the CPU
context.

Commit:
Brought back eager based runtime, as a stepping stone towards tensor handle support.

Switched CPU and GPU runtime unit tests to eager.

This PR adapts and extends the deleted code from https://github.com/apple/swift/commit/684177fe94a1eb7a03f4f1403cd0473a70f885bc#diff-3d9a6452da404ba617a54693692a94f9.

Disabled a few unit tests for GPU execution, and will file bugs to track
them. This PR is however expected to be usable for Jupyter notebook in the CPU
context.

Commit:
Added design notes as a comment block to CompilerRuntime.swift.

mhong commented:
@swift-ci please clean test tensorflow

mhong commented:
@alextp as FYI

rxwei commented on stdlib/public/TensorFlow/TensorHandle.swift:
```diff
@@ -121,14 +131,21 @@ extension TensorHandle : TensorSendableReceivable {
+    if _RuntimeConfig.usesTFEagerAPI {
+      let context = _ExecutionContext.global
+      TFE_EnqueueNamedTensorFromCtx(
+        context.eagerCtx, Int32(tensorId), self.cTensorHandle, status)
```
`tensorId` -> `tensorID`

mhong replied:
Changed, but this is existing code, and I feel `tensorId` would also be fine.

rxwei replied:
In Swift readability reviews we tend to prefer "ID", but it's undocumented in Google's swift style guide. The way I reason about it is as follows:
- "Id" is a shortened version of "identifier".
- "ID" is fine and does not need to be considered a shortened name, because in English "ID" is a word.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -273,15 +296,15 @@ public final class _ExecutionContext {
     let ctx = TFE_NewContext(opts, status)
     checkOk(status)
-    self.cContext = ctx!
+    self.eagerCtx = ctx!
```
`eagerCtx` -> `eagerContext`

mhong replied:
Changed, but from similar past discussion, i feel always spelling out the full word may not add additional clarity. ctx is a well-understood term.

rxwei replied:
This is not a question of clarity, but a question of naming convention. I tried to grep `ctx` in `stdlib/core/` and I found none. Plus, the standard library implementation is consistently following the naming convention by spelling out words for type members, functions and types AFAIK. It's better that we strive to do the same.

mhong replied:
I still think this kind of convention is hard to maintain in an absolute way. Say we have a class named `MatrixMultiplicationManager` -- I think it would be clearer if it's instead sth like `MatmulMgr`. So where do we draw the line on when to stop using full words?

stdlib is an evolving entity. If someday it adds more systems-oriented libraries (e.g. for managing thread context, context switch, etc), I won't be surprised if some authors go for `ctx`.

It'd be better if we can relax our coding convention a bit on this subject, and leave it to personal choices (unless the chosen name reduces clarity).

rxwei replied:
> Say we have a class named `MatrixMultiplicationManager` -- I think it would be clearer if it's instead sth like `MatmulMgr`.

"Manager" and "Context" are really common words across Swift code bases, and the conventional spelling of these two words _are_ the full form.

I have never seen anyone use "Mgr" in Swift on a type name. Plus, "Mgr" would _definitely not_ be considered acceptable as a _public_ API. If you are talking about internal API, these things are hard to argue about. In these case, I would choose not to argue but just refer to [Swift API Design Guidelines - Section "Use Terminology Well" - Item "Avoid abbreviations"](https://swift.org/documentation/api-design-guidelines).

>  If someday it adds more systems-oriented libraries (e.g. for managing thread context, context switch, etc), I won't be surprised if some authors go for `ctx`.

That would not surprise me either, only if it were local variables or private code.

I don't think the argument of "more system-oriented libraries" would make the case more convincing. For example, there's hardly anything more low-level than pointer types, but they are called `UnsafeMutableBufferPointer` instead of `UnsafeMutableBufPtr`. The reason for that is "Avoid abbreviations" defined in the Swift API design guidelines.

But again, if it's internal/private code, it's hard to argue about. However, in Swift readability reviews, we tend to treat external and internal code the same.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -638,14 +767,32 @@ public final class _TensorComputation {
+  private var pthreads: [pthread_t] = []
+
+  /// The data structure to pass into pthread creation API. 
+  class FuncHandle {
```
I don't quite understand why a separate class is needed. Since now every `_TensorComputation` is getting a thread index (unless its sync runtime of course), can we simply add a `var threadIndex: Int?` to `_TensorComputation` to avoid an additional level of complication?

mhong replied:
Added a comment: We cannot have the ThreadBody closure below close over `threadIndex`, because ThreadBody is of C convention.

rxwei replied:
I understand that this is to work around that this is a C convention function. My question was: Why not just add the index as a member in `_TensorComputation`? Then you would not need to change the original self-passing code at all.

mhong replied:
We are calling `_TensorComputation.execute()` in concurrent threads, so if we make `threadIndex` a property, we'll need to mutex-protect it. I think this change here is simpler.

rxwei replied:
I see! Would it be possible to have a clearer name? I don't quite get what "Func" is referring to. Something like `_TensorComputationThreadHandle`?

mhong replied:
`_TensorComputationThreadHandle` is a bit too long. :-)

I settled on `ThreadParam` -- it's local to `_TensorComputation`, so the context should make its semantics clear. WDYT?

rxwei replied:
Sure thing!

rxwei replied:
Another minor nit: the class can be `private`.

mhong replied:
SG. will address in the next PR.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -638,14 +767,32 @@ public final class _TensorComputation {
+  /// The data structure to pass into pthread creation API. 
+  class FuncHandle {
+    public let computation: _TensorComputation
+    public let threadIdx: Int
```
`threadIdx` -> `threadIndex`

mhong replied:
Done, though I don't think this adds clarity, similar to the above thread.

rxwei commented on stdlib/public/TensorFlow/TensorHandle.swift:
```diff
@@ -121,14 +131,21 @@ extension TensorHandle : TensorSendableReceivable {
+    if _RuntimeConfig.usesTFEagerAPI {
+      let context = _ExecutionContext.global
+      TFE_EnqueueNamedTensorFromCtx(
+        context.eagerCtx, Int32(tensorId), self.cTensorHandle, status)
```
`self.` is unnecessary.

mhong replied:
Removed, but I think having `self.` helps distinguish properties from local vars. WDTY?

rxwei replied:
I don't have an opinion other than a very slight preference toward consistency across the TensorFlow module. I agree that exceptions can occur when reasonable.

rxwei commented on stdlib/public/TensorFlow/TensorHandle.swift:
```diff
@@ -121,14 +131,21 @@ extension TensorHandle : TensorSendableReceivable {
+      TFE_EnqueueNamedTensorFromCtx(
+        context.eagerCtx, Int32(tensorId), self.cTensorHandle, status)
+    } else {
+      let cTensor = TFE_TensorHandleResolve(self.cTensorHandle, status)
```
`self.` is unnecessary. Though it doesn't matter, we've been consistently avoiding unnecessary `self.` so it'd be nice to maintain that consistency.

mhong replied:
Done.

rxwei commented on stdlib/public/TensorFlow/TensorHandle.swift:
```diff
@@ -148,16 +165,12 @@ internal extension ShapedArray where Scalar : AccelerableByTensorFlow {
-      return ret!
-    }
+    debugLog("Calling TFE_TensorHandleCopyToDevice().")
+    let hostHandle: CTensorHandle? = TFE_TensorHandleCopyToDevice(
```
The type signature here seems redundant. But if you prefer to keep it for clarity that's reasonable too!

mhong replied:
Yes i feel it adds clarity.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -767,14 +937,41 @@ public final class _TensorComputation {
   // NOTE: This is to be called by the initializer. The computation gets
   // executed on initialization, thus this method will not be exposed to users.
-  private func execute() {
+  private func execute(threadIdx: Int) {
```
`threadIndex`

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -767,14 +937,41 @@ public final class _TensorComputation {
-  /// Runs the tensor program. Aborts the process on error, and emits an error
-  /// string to STDERR.
+  /// In eager mode, runs a device function in a corresponding thread given by
+  /// `threadIdx`. Otherwise, runs all devices functions of the tensor
```
`threadIndex`

mhong replied:
Done

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -767,14 +937,41 @@ public final class _TensorComputation {
+        var returnValueCount = Int32(returnValues.count)
+        TFE_Execute(op, &returnValues, &returnValueCount, status)
+        debugLog("""
+                   returnValues.count=\(returnValues.count), \
```
Instead of indenting all the way through `"""` like this, we've been using a different style for multi-line string literals:

```swift
debugLog("""
  returnValues.count=\(returnValues.count), \
  returnValueCount=\(returnValueCount).
  """)
```

mhong replied:
Changed, though I should note that my emacs auto indentation picked the former. So it'd be nice if we don't have to "fight the tool".

rxwei replied:
Well, emacs doesn't really enforce the standard Swift format. I doubt emacs' swift-mode even has support for multi-line string literals in Swift, which were introduced not too long ago.

If the tool was `swift-format` or `swiftlint`, then I would definite agree that not fighting the tool is ideal. Hopefully they will be available soon.

mhong replied:
SG

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -767,14 +937,41 @@ public final class _TensorComputation {
+                   returnValues.count=\(returnValues.count), \
+                   returnValueCount=\(returnValueCount).
+                   """)
+        assert(returnValueCount == returnValues.count)
```
Use `internalConsistencyCheck` for consistency with line 962.

mhong replied:
Good catch! Done.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -638,14 +767,32 @@ public final class _TensorComputation {
+  ///
+  /// The global config flag '_RuntimeConfig.usesSynchronousExecution' decides
+  /// whether tensor computation should be synchronous: if true, this property
+  /// will always be nil. Otherwise, this property is non-nil only when the
```
"this property will always be nil" -> "this property will always be empty". "Nil" refers exclusively to `Optional.none` in Swift.

mhong replied:
You are right. changed.

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -638,14 +767,32 @@ public final class _TensorComputation {
+  ///
+  /// The global config flag '_RuntimeConfig.usesSynchronousExecution' decides
+  /// whether tensor computation should be synchronous: if true, this property
+  /// will always be nil. Otherwise, this property is non-nil only when the
```
"this property is non-nil" -> "this property is non-empty"

mhong replied:
Done.

rxwei commented on stdlib/public/TensorFlow/TensorHandle.swift:
```diff
@@ -97,15 +97,25 @@ extension TensorHandle : TensorSendableReceivable {
+      checkOk(status)
+      tensorHandle = TensorHandle<Scalar>(owning: cTensorHandle!)
+    } else {
+      let cTensor: CTensor? = TF_DequeueNamedTensor(
```
This forms a perfect use case for implicitly unwrapped optionals. You can do
```swift
let cTensor: CTensor! = ...
checkOk(status)
internalConsistencyCheck(cTensor != nil, ...)
tensorHandle = TensorHandle<Scalar>(copyingFromCTensor: cTensor) // no ! needed here and below
TF_DeleteTensor(cTensor)
```

Since this is just some old code being moved around, it's fine too keep it as is!

mhong replied:
hmm, if TF_DequeueNamedTensor() returns NULL, wouldn't `let cTensor: CTensor! = ...` crash?

rxwei replied:
It would not crash, because IUOs are only unwrapped when you pass them around as a non-optional. `let cTensor: CTensor!` is just an initialization, not passing.

In other words, IUOs are just optionals that get implicitly unwrapped when you use them in positions where the contextual type is not optional.

mhong replied:
Thanks for clarifying. I changed the code to use IUOs. The code is indeed simpler!

Commit:
Addressed comments from @rxwei.

mhong commented:
Thanks for the review! I addressed most of the comments (for some of them, I also expressed alternative opinions above, and would be great to get your input). 

P.S. As I reread my reply comments above, they might come across as not being sufficiently open to review feedback. That's not the case. Although some of the style discussion can be a matter of person choice, I'd really love to understand more about your swift style and the associated merits!

mhong commented:
CI tests passed:
https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux/794/
https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-Linux-gpu/76/ (Swift(linux-x86_64).SourceKit/Refactoring.basic.swift appears is flaky)

Commit:
Addressed more comments.

rxwei reviewed (approved):
Great progress!

rxwei commented:
> As I reread my reply comments above, they might come across as not being sufficiently open to review feedback

Hey, I didn't feel that way at all! And I do agree your personal choices on, e.g. using `self.`, make sense!

mhong commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -638,14 +767,34 @@ public final class _TensorComputation {
+  /// We cannot have the ThreadBody closure below close over on `threadIndex`,
+  /// because ThreadBody is of C convention.
+  class FuncHandle {
+    public let computation: _TensorComputation
```
Good catch. removed `public` here. Also removed the initializer as you suggested. Thanks!

mhong commented on test/TensorFlowRuntime/dataset_api.swift:
```diff
@@ -15,6 +15,10 @@ import StdlibUnittest
 
+// TODO: fix the test suite for GPU.
+// We have errors like:
+// Not found: No registered 'ZipDataset' OpKernel for GPU devices compatible with node {{node op/_S4mainyycfU4_.tf_GPU.70.64_/device_GPU_0_7}} = ZipDataset[N=2, output_shapes=[<unknown>, <unknown>], output_types=[DT_FLOAT, DT_FLOAT], _device="/device:GPU:0"](tf_recv_55, tf_recv_56)
```
Yes that sounds like it. I added this thread to https://bugs.swift.org/browse/SR-8711. Thanks!

mhong commented:
Thanks again for the review comments and quick turn-around!

I'm submitting this PR to make progress. Happy to address additional comments if any.

mhong commented:
Update: I did not get to address the last comment in the submitted PR. Here's a follow-up.

When we remove the initializer for `ThreadParam`, I got compiler errors:

```
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:789:9: error: class '_TensorComputation.ThreadParam' has no initializers
  class ThreadParam {
        ^
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:790:9: note: stored property 'computation' without initial value prevents synthesized initializers
    let computation: _TensorComputation
        ^
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:791:9: note: stored property 'threadIndex' without initial value prevents synthesized initializers
    let threadIndex: Int
        ^
                         = 0
/usr/local/google/home/hongm/ssd_part/git/swift-base/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:913:13: error: '_TensorComputation.ThreadParam' cannot be constructed because it has no accessible initializers
            ThreadParam(computation: self,
            ^
```

rxwei commented on stdlib/public/TensorFlow/CompilerRuntime.swift:
```diff
@@ -452,6 +514,73 @@ internal func dumpCTensorHandleContent(
+    let entryFunctionBaseName = String(cString: entryFunctionBaseNameAddress)
+    debugLog("Looking up op(s) from func base name \(entryFunctionBaseName).")
+    for i in 0...helperFunctionCount {
+      /// Also look up in the TensorFlow program a function name (op type) based
```
Not doc comments, so double slash instead of triple :)

rxwei commented:
> When I remove the initializer for `ThreadParam`, I got compiler errors:

Interesting... Ok!

lattner commented:
Nice progress!!

rxwei commented:
This patch broke macOS builds due to inconsistency in pthread APIs.

```swift
/Users/rxwei/Development/Swift/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:923:25: error: value of optional type 'pthread_t?' (aka 'Optional<UnsafeMutablePointer<_opaque_pthread_t>>') must be unwrapped to a value of type 'pthread_t' (aka 'UnsafeMutablePointer<_opaque_pthread_t>')
        pthreads.append(newThread)
                        ^
/Users/rxwei/Development/Swift/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:923:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
        pthreads.append(newThread)
                        ^
                                  ?? <#default value#>
/Users/rxwei/Development/Swift/swift/stdlib/public/TensorFlow/CompilerRuntime.swift:923:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
        pthreads.append(newThread)
                        ^
                                 !
```

Next time we should make sure macOS tests pass when low-level logic is changed (e.g. pthreads).

mhong commented:
Agree. For the Mac CI run of this PR, I found a different error: https://ci-external.swift.org/view/Pull%20Request/job/swift-PR-TensorFlow-macOS/51/console. Is that a known CI breakage?

rxwei commented:
That's unknown. macOS CI has not been properly set up yet. I'll follow up with @shahmishal.
